sympy/ntheory/ecm.py

```diff
@@ -0,0 +1,147 @@
+class point:
+    """For denoting the point on the curve.
+    For ECM we don't need the y-coordinate. All
+    the arthematic are done with x and z coordinate"""
+    def __init__(self, x, z):
+        self.x = x
+        self.z = z
+
+def _pt_add(p1, p2, diff, n):
+    """Returns p1 + p2.
+
+    Parameters
+    ==========
+
+    p1 : point on the curve
+    p2 : point on the curve
+    diff : p1 - p2
+    n : Modulus
+    """
+    a = (p1.x - p1.z) * (p2.x + p2.z)
+    b = (p1.x + p1.z) * (p2.x - p2.z)
+    x_cor = diff.z*((a + b)**2)
+    z_cor = diff.x*((a - b)**2)
+    return point(x_cor%n, z_cor%n)
+
+def _pt_double(p1, A, n):
+    """
+    For doubling of the points on the curve.
+    """
+    a = (p1.x + p1.z)**2%n
+    b = (p1.x - p1.z)**2%n
+    An, Ad = A
+    ab = a - b
+    x_cor = 4*a*b*Ad % n
+    z_cor = (4*Ad*b + ab*An)*ab % n
+    return point(x_cor%n, z_cor%n)
+
+def _mont_ladder(p_0, k, A, n):
+    """Scalar multiplication of points done
+    using Montgomery Ladder algorithm.
+
+    Parameters
+    ==========
+    p_0 : point on the curve
+    k : Integer
+    A : parameter of the curve
+    n : Modulus
+    """
+    k = bin(k)[3:]
+    q = p_0
+    p = _pt_double(p_0, A, n)
+    for i in k:
+        if i == '1':
+            q = _pt_add(p, q, p_0, n)
+            p = _pt_double(p, A, n)
+        else:
+            p = _pt_add(q, p, p_0, n)
+            q = _pt_double(q, A, n)
+    return q
+
+def ilog(x, b):
+    """Returns least integer l for which b**l >= x"""
+    l = 0
+    while x >= b:
+        x /= b
+        l += 1
+    return l
+
+
+def ecm(n, B1=100, B2=200, max_curve=5):
+    """Lenstra's Elliptic curve Factorization using
+    2 stage with Suyuma's Parameterization.
+
+    References
+    ==========
+
+    .. [1]  Carl Pomerance and Richard Crandall "Prime Numbers:
+        A Computational Perspective" (2nd Ed.), page 344
+    """
+    from sympy.ntheory import prime, sieve, isprime
+    from sympy import gcd
+    import random
```

Comment:
Look at other routines that use random; this will need a `seed` parameter so the output can be made canonical and testable.